aequilibrae/paths/select_link.py

```diff
@@ -0,0 +1,164 @@
+        """ Linear approximation contribution for each iteration. """
+
+        # solution^n+1 = alpha^n * sol^n + (1-alpha^n) * direction^n
+        # direction = beta_0 * aon + beta_1 * previous_direction + beta_2 * pre_previous_direction
+        alphas = self.paths.assignment_results.get_alphas()
+        print(alphas)
+        # betas_0 = assignment_report["convergence"]["beta0"]
+        # betas_1 = assignment_report["convergence"]["beta1"]
+        # betas_2 = assignment_report["convergence"]["beta2"]
+
+        self.demand_weights = np.repeat(1.0, self.num_iters)
+
+        # FIXME (Jan 21/4/21): implement CFW and BFW, this is only valid for FW
+        # can we just multiply previous like below?
+        # if assignment_report["setup"]["Algorithm"] in ["fw", "frank-wolfe"]:
+        for i in range(0, self.num_iters):
+            alpha = alphas[i]
+            # beta_0 = betas_0[i]
+            # beta_1 = betas_1[i]
+            # beta_2 = betas_2[i]
+            # demand_weights[i] += alpha * direction + (1.0 - alpha) * current
+            # for FW: direction = AON, so multiply current weight
+            # for CFW: direction = beta_0 * AON + beta_1 * previous_direction
+            # for BFW: direction = beta_0 * AON + beta_1 * previous_direction + beta_2 * pre_previous_direction
+            self.demand_weights[i] *= alpha
+            self.demand_weights[:i] *= 1.0 - alpha
+
+            # if (assignment_report["setup"]["Algorithm"] in ["cfw", "bfw"]) and (i > 1):
+            #     self.demand_weights[i] *= beta_0
+            #     self.demand_weights[:(i - 1)] *= beta_1
+            #
+            # if (assignment_report["setup"]["Algorithm"] == "bfw") and (i > 2):
+            #     self.demand_weights[:(i-2)] *= beta_2
+
+    def __lookup_compressed_links_for_link(self, link_ids: List[int]) -> List[int]:
+        """" TODO: look up compressed ids for each class for a given list of network link ids"""
+        # FIXME: for now just pass in simplified ids directly
+        select_link_ids_compressed = {}
+        for c in self.classes:
+            select_link_ids_compressed[c.__id__] = link_ids
+        #     graph = self.paths.compressed_graph_correspondences[c.__id__]
+        #     select_link_ids_compressed[c.__id__] = graph.loc[graph["link_id"].isin(link_ids)][
+        #         "__compressed_id__"
+        #     ].to_numpy()
+        return select_link_ids_compressed
+
+    def __initialise_matrices(self, simplified_link_ids: List[int]) -> Dict[str, Dict[int, np.array]]:
+        """ For each class and each link, initialise select link demand matrix"""
+        select_link_matrices = {
+            c.__id__: {link_id: np.zeros_like(self.demand_matrices[c.__id__].matrix_view)}
+            for c in self.classes
+            for link_id in simplified_link_ids[c.__id__]
+        }
+        return select_link_matrices
+
+    def run_select_link_analysis(self, link_ids: List[int]) -> None:
```

Comment:
Should the annotation be " --> AequilibraeMatrix"?

Reply:
it's a dict of dict (mode and link id) of np.array, but should probably make it an Aequlibraematrix  instead of np.array